apps/emqx_gateway/src/coap/emqx_coap_api.erl

```diff
@@ -18,93 +18,89 @@
 
 -behaviour(minirest_api).
 
+-include_lib("typerefl/include/types.hrl").
+-include_lib("emqx/include/logger.hrl").
 -include("src/coap/include/emqx_coap.hrl").
 
 %% API
--export([api_spec/0]).
+-export([api_spec/0, paths/0, schema/1, namespace/0]).
 
 -export([request/2]).
 
 -define(PREFIX, "/gateway/coap/clients/:clientid").
--define(DEF_WAIT_TIME, 10).
 
--import(emqx_mgmt_util, [ schema/1
-                        , schema/2
-                        , object_schema/1
-                        , object_schema/2
-                        , error_schema/2
-                        , properties/1]).
+-import(hoconsc, [mk/2, enum/1]).
+-import(emqx_dashboard_swagger, [error_codes/2]).
 
 %%--------------------------------------------------------------------
 %%  API
 %%--------------------------------------------------------------------
+namespace() -> "gateway_coap".
+
 api_spec() ->
-    {[request_api()], []}.
+    emqx_dashboard_swagger:spec(?MODULE, #{check_schema => true, translate_body => true}).
+
+paths() ->
+    [?PREFIX ++ "/request"].
+
+schema(?PREFIX ++ "/request") ->
+    #{operationId => request,
+      post => #{ tags => [<<"gateway|coap">>]
+               , description => <<"Send a CoAP request message to the client">>
+               , parameters => request_parameters()
+               , requestBody => request_body()
+               , responses => #{200 => coap_message(),
+                                404 => error_codes(['CLIENT_NOT_FOUND'], <<"client not found error">>),
+                                504 => error_codes(['CLIENT_NOT_RESPONSE'], <<"timeout">>)}
+               }
+     }.
 
-request_api() ->
-    Metadata = #{post => request_method_meta()},
-    {?PREFIX ++ "/request", Metadata, request}.
 
 request(post, #{body := Body, bindings := Bindings}) ->
     ClientId = maps:get(clientid, Bindings, undefined),
-
-    Method = maps:get(<<"method">>, Body, <<"get">>),
-    CT = maps:get(<<"content_type">>, Body, <<"text/plain">>),
+    Method = maps:get(<<"method">>, Body, get),
+    AtomCT = maps:get(<<"content_type">>, Body),
     Token = maps:get(<<"token">>, Body, <<>>),
```

Comment:
> The token value is a sequence of 0 to 8 bytes.

From the protocol spec, Token should have a length limit, we can keep improving later.